File: backfire-kernel/crates/backfire-core/src/knowledge.rs
// ─────────────────────────────────────────────────────────────────────
// Director-Class AI — Ground Truth Store (RAG Interface)
// ─────────────────────────────────────────────────────────────────────
//! RAG (Retrieval-Augmented Generation) interface for fact retrieval.
//!
//! The in-memory backend provides keyword matching for testing and
//! low-latency (<1ms) retrieval. Production deployments can plug in
//! ChromaDB or any vector DB via the `GroundTruthStore` trait.

use std::collections::HashMap;

/// Trait for ground truth retrieval backends.
pub trait GroundTruthStore: Send + Sync {
    /// Retrieve relevant context for the given query.
    /// Returns `None` if no relevant facts are found.
    fn retrieve_context(&self, query: &str) -> Option<String>;
}

/// In-memory keyword-based ground truth store.
///
/// Mirrors `GroundTruthStore` from `knowledge.py:12-56`.
pub struct InMemoryKnowledge {
    facts: HashMap<String, String>,
}

impl Default for InMemoryKnowledge {
    fn default() -> Self {
        let mut facts = HashMap::new();
        facts.insert("sky color".into(), "blue".into());
        facts.insert("system layers".into(), "16".into());
        facts.insert("sec metric".into(), "sustainable ethical coherence".into());
        facts.insert("backfire limit".into(), "entropy threshold".into());
        Self { facts }
    }
}

impl InMemoryKnowledge {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_facts(facts: HashMap<String, String>) -> Self {
        Self { facts }
    }

    pub fn add_fact(&mut self, key: String, value: String) {
        self.facts.insert(key, value);
    }
}

impl GroundTruthStore for InMemoryKnowledge {
    fn retrieve_context(&self, query: &str) -> Option<String> {
        let query_lower = query.to_lowercase();
        let mut context = Vec::new();

        for (key, value) in &self.facts {
            let key_words: Vec<&str> = key.split_whitespace().collect();
            if key_words.iter().any(|w| query_lower.contains(w)) {
                context.push(format!("{key} is {value}"));
            }
        }

        if context.is_empty() {
            None
        } else {
            Some(context.join("; "))
        }
    }
}

/// External ground truth store that calls a function pointer.
///
/// Used by the PyO3 FFI layer to delegate RAG retrieval to Python.
type RetrieveFn = Box<dyn Fn(&str) -> Option<String> + Send + Sync>;

pub struct ExternalKnowledge {
    retrieve_fn: RetrieveFn,
}

impl ExternalKnowledge {
    pub fn new(retrieve_fn: impl Fn(&str) -> Option<String> + Send + Sync + 'static) -> Self {
        Self {
            retrieve_fn: Box::new(retrieve_fn),
        }
    }
}

impl GroundTruthStore for ExternalKnowledge {
    fn retrieve_context(&self, query: &str) -> Option<String> {
        (self.retrieve_fn)(query)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_facts() {
        let store = InMemoryKnowledge::new();
        let ctx = store.retrieve_context("What color is the sky?");
        assert!(ctx.is_some());
        assert!(ctx.unwrap().contains("blue"));
    }

    #[test]
    fn test_system_layers() {
        let store = InMemoryKnowledge::new();
        let ctx = store.retrieve_context("How many system layers?");
        assert!(ctx.is_some());
        assert!(ctx.unwrap().contains("16"));
    }

    #[test]
    fn test_truly_no_match() {
        let store = InMemoryKnowledge::new();
        let ctx = store.retrieve_context("xyzzy plugh");
        assert!(ctx.is_none());
    }

    #[test]
    fn test_custom_facts() {
        let mut facts = HashMap::new();
        facts.insert("test key".into(), "test value".into());
        let store = InMemoryKnowledge::with_facts(facts);
        let ctx = store.retrieve_context("test query with key");
        assert!(ctx.is_some());
        assert!(ctx.unwrap().contains("test value"));
    }

    #[test]
    fn test_external_knowledge() {
        let store = ExternalKnowledge::new(|q| {
            if q.contains("sky") {
                Some("sky is blue".into())
            } else {
                None
            }
        });
        assert!(store.retrieve_context("sky color").is_some());
        assert!(store.retrieve_context("unrelated").is_none());
    }

    #[test]
    fn test_add_fact() {
        let mut store = InMemoryKnowledge::with_facts(HashMap::new());
        assert!(store.retrieve_context("capital").is_none());
        store.add_fact("capital".into(), "Paris".into());
        let ctx = store.retrieve_context("What is the capital?");
        assert!(ctx.is_some());
        assert!(ctx.unwrap().contains("Paris"));
    }

    #[test]
    fn test_add_fact_overwrites() {
        let mut store = InMemoryKnowledge::new();
        store.add_fact("sky color".into(), "green".into());
        let ctx = store.retrieve_context("sky color").unwrap();
        assert!(ctx.contains("green"));
        assert!(!ctx.contains("blue"));
    }
}

File: src/director_ai/core/retrieval/knowledge.py
# SPDX-License-Identifier: Apache-2.0
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# Director-Class AI — Ground Truth Store (RAG Interface)

"""Ground-truth knowledge store: the RAG interface backing grounded answers."""

from __future__ import annotations

import hashlib
import logging
from typing import TYPE_CHECKING, Any

from ..text_overlap import word_overlap

if TYPE_CHECKING:
    from ..types import EvidenceChunk

__all__ = ["GroundTruthStore"]

# Common English function words filtered before value-content matching, so a
# query never matches a fact on stop words alone. Deliberately small and
# recall-first — this only gates the degraded keyword fallback, not the primary
# vector path.
_STOPWORDS = frozenset(
    {
        "a",
        "an",
        "and",
        "are",
        "as",
        "at",
        "be",
        "by",
        "do",
        "does",
        "did",
        "for",
        "from",
        "has",
        "have",
        "how",
        "i",
        "in",
        "is",
        "it",
        "its",
        "many",
        "me",
        "my",
        "of",
        "on",
        "or",
        "that",
        "the",
        "this",
        "to",
        "was",
        "what",
        "when",
        "where",
        "which",
        "who",
        "why",
        "will",
        "with",
        "you",
        "your",
    }
)


class GroundTruthStore:
    """In-memory fact store with keyword matching retrieval.

    Stores keyâ†’value fact pairs and retrieves by word overlap between
    the query and fact keys. No embeddings or semantic similarity.

    For vector-based retrieval, use ``VectorGroundTruthStore`` with a
    ``VectorBackend`` (Chroma, Pinecone, Qdrant, FAISS, etc.).
    """

    _DEMO_FACTS = {
        "sky color": "blue",
        "scpn layers": "16",
        "layer 1": "quantum biological",
        "layer 16": "director",
        "sec metric": "sustainable ethical coherence",
        "backfire limit": "entropy threshold",
        "vibrana symmetry": "13-fold",
    }

    def __init__(self) -> None:
        self.logger = logging.getLogger("DirectorAI.GroundTruthStore")
        self.facts: dict[str, str] = {}
        self._revision = 0

    @classmethod
    def with_demo_facts(cls) -> GroundTruthStore:
        """Return a store pre-loaded with demo facts (for tests and --demo)."""
        store = cls()
        store.facts.update(cls._DEMO_FACTS)
        return store

    def add(
        self,
        key: str,
        value: str,
        metadata: dict[str, Any] | None = None,
        tenant_id: str = "",
    ) -> None:
        """Add or update a fact in the store.

        ``metadata`` is accepted for Liskov compatibility with
        :class:`VectorGroundTruthStore.add` but ignored here — the
        keyword store keeps no structured metadata.
        """
        _ = metadata  # intentional no-op for LSP compat
        key = _require_non_empty_string("key", key)
        value = _require_non_empty_string("value", value)
        tenant_id = tenant_id.strip()
        full_key = f"{tenant_id}:{key}" if tenant_id else key
        self.facts[full_key] = value
        self._bump_revision()

    def add_fact(self, key: str, value: str, tenant_id: str = "") -> None:
        """Alias for add() — used by some callers."""
        self.add(key, value, tenant_id=tenant_id)

    @property
    def revision(self) -> int:
        """Monotonic store revision used to scope score-cache entries."""
        return self._revision

    def _bump_revision(self) -> None:
        self._revision += 1

    def cache_scope(self, tenant_id: str = "") -> str:
        """Return a stable cache scope for facts visible to a tenant."""
        return f"{type(self).__name__}:tenant={tenant_id}:revision={self._revision}"

    def retrieve_context_with_chunks(
        self,
        query: str,
        top_k: int = 3,
        tenant_id: str = "",
    ) -> list[EvidenceChunk]:
        """Return the keyword-retrieved context wrapped as evidence chunks."""
        from ..types import EvidenceChunk

        context_str = self.retrieve_context(query, tenant_id=tenant_id)
        if not context_str:
            return []
        return [EvidenceChunk(text=context_str, distance=0.0, source="keyword")]

    def retrieve_context(
        self, query: str, top_k: int = 0, tenant_id: str = ""
    ) -> str | None:
        """Retrieve relevant facts matching *query*.

        Returns a semicolon-separated context string, or ``None`` if
        no relevant facts are found.
        """
        query = _require_non_empty_string("query", query)
        tenant_id = tenant_id.strip()
        if top_k < 0:
            raise ValueError(f"top_k must be >= 0; got {top_k!r}")
        if not self.facts:
            self.logger.info(
                "GroundTruthStore is empty — add facts via .add() "
                "or use VectorGroundTruthStore.ingest()",
            )
            return None

        query_lower = query.lower()
        query_tokens = {w for w in query_lower.split() if w not in _STOPWORDS}
        ranked: list[tuple[float, str]] = []

        for key, value in self.facts.items():
            search_key = key
            if tenant_id:
                prefix = f"{tenant_id}:"
                if not key.startswith(prefix):
                    continue
                search_key = key[len(prefix) :]
            key_words = search_key.lower().split()
            if any(word in query_lower for word in key_words):
                # Curated, semantically-keyed fact: rank on the key (unchanged).
                ranked.append((_word_overlap(query, search_key), value))
                continue
            # Value-aware fallback for content-keyed facts — e.g. API-ingested
            # chunks keyed by an opaque cid ("doc:chunk:0"), which has no query
            # words, so key matching never reaches the chunk text. Match the
            # value, stopword-filtered so a query cannot match on function words
            # alone, and rank below key hits so curated facts always win.
            value_tokens = {w for w in value.lower().split() if w not in _STOPWORDS}
            if query_tokens & value_tokens:
                ranked.append((0.5 * _word_overlap(query, value), value))

        if ranked:
            ranked.sort(key=lambda item: item[0], reverse=True)
            context = [value for _, value in ranked]
            if top_k > 0:
                context = context[:top_k]
            retrieved = "; ".join(context)
            qhash = hashlib.sha256(query.encode()).hexdigest()[:12]
            self.logger.info(
                "RAG Retrieval: %d facts matched (query=%s, len=%d)",
                len(context),
                qhash,
                len(query),
            )
            return retrieved

        return None


def _require_non_empty_string(name: str, value: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{name} must be a non-empty string")
    return value.strip()


def _word_overlap(text_a: str, text_b: str) -> float:
    """Return lexical Jaccard overlap in ``[0, 1]``.

    Delegates to the shared measured-fast-path helper (pure Python below a large
    -input threshold, Rust above it). See :mod:`director_ai.core.text_overlap`.
    """
    return word_overlap(text_a, text_b, logger_name=__name__)
